google/cloud/apigee/registry/applications/v1alpha1/registry_styleguide.proto

```diff
@@ -0,0 +1,121 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package google.cloud.apigee.registry.applications.v1alpha1;
+
+import "google/api/field_behavior.proto";
+
+option java_package = "com.google.cloud.apigee.registry.applications.v1alpha1";
+option java_multiple_files = true;
+option java_outer_classname = "RegistryStyleGuideProto";
+option go_package = "github.com/apigee/registry/rpc;rpc";
+
+// StyleGuide defines a set of guidelines, rules, and linters that govern
+// the static analysis of API Specs.
+message StyleGuide {
+    // Resource name of the style guide.
+    string name = 1 [
+        (google.api.field_behavior) = REQUIRED
+    ];
+
+    // A list containing style (format) descriptors for this style guide that 
+    // are specified as a Media Type (https://en.wikipedia.org/wiki/Media_type).
+    // Possible values include "application/vnd.apigee.proto", 
+    // "application/vnd.apigee.openapi", and "application/vnd.apigee.graphql",
+    // with possible suffixes representing compression types. These hypothetical
+    // names are defined in the vendor tree defined in RFC6838 
+    // (https://tools.ietf.org/html/rfc6838) and are not final.
+    repeated string mime_types = 2 [
+		(google.api.field_behavior) = REQUIRED
+	];
```

Comment:
Why multiple MIME types? It's unclear to me from this description how I would use multiple types for one style guide. Do you have an example or something that can be added to the description to clarify?

Reply:
In the short term this allows us to match variations in mime type strings related to compression or OpenAPI version without someone having to write a general mime type matching service.

In the long term this allows for a style guide to support specs of different formats - we expect but wouldn't require all style guides to apply to exactly one format - this would really depend on what the linters being used would accept. We also considered that an official mime type for OpenAPI (outside of the vendor space) might appear and that a registry might include a mixture of these types (messy but not the messiest thing that has ever happened)

Reply:
You've lost me from the first sentence because I'm failing to understand how the field is used. I'll have to revisit your points after you help me get even a basic grasp on what this field does. An example would be really helpful. What does it mean if I create a style guide with the following MIME types?

```
mime_types = [
    "application/vnd.apigee.openapi",
    "application/vnd.apigee.graphql",
]
```

I currently have no idea what that represents, and I've tried my best to figure it out given the documentation. That said, there's almost certainly room for improvement in this field's comment. If I can't even figure this out, who is the intended audience of this documentation? What are we assuming they understand a priori?

>A list containing style (format) descriptors for this style guide that are specified as a Media Type

I'll call out this sentence specifically. What does it mean? How can a style guide have multiple formats simultaneously? Unless the formats are overlapping (e.g. JSON and YAML) I'm not seeing how this makes any sense.

Any corrections on my misunderstandings or answers to my questions would be appreciated.

Reply:
@seaneganx I agree that the documentation can and will be improved in a subsequent PR. Here's an example of what Tim was talking about:

```yaml
mime_types:
   - application/x.openapi+gzip;version=3.0.0
   - application/x.openapi+gzip;version=2
```

As in, "the rules in this style guide apply to OpenAPI Version 2 and Version 3 specs". In the future, the registry may support mime type parsing (where we can parse out the message, compression, version and this can be refactored, and so the user would simply write:

```yaml
mime_types:
   - application/x.openapi
```

Until then, we're choosing to allow multiple mime types. I'll make this explicit and clear in a subsequent PR that I'll open up, because the documentation does not reflect this currently.

Reply:
Thanks, that clears everything up. This field lists the MIME types of the specs that the style guide applies to.

Without that in the documentation I misunderstood it as the MIME types of the style guide itself since it's a field on the style guide resource. Our other APIs have MIME type fields that indicate that resource's type, so we will want to make this difference clear.

Now for a subtle detail. Are rules generic or do they only apply to specific MIME types?

Reply:
The short answer is yes. Each style guide applies to a list of given MIME types. The style guide has a list of guidelines, which contain rules. For now, it is the user's job to make sure that the rules that they provide in the guidelines of a style guide actually apply to the mime type that is specified in the style guide. When this is actually being implemented (and I get to the point where I'm implementing error-handling), I will implement a way to make sure that the rules that the user is applying to the MIME type are actually valid for that specific MIME type.

Reply:
We should keep in mind that the style guide YAML will serve as a user interface for people who are writing and configuring style guides, so we probably want the format to be as clear and unsurprising as possible.

Subjectively, I don't expect that we would see many style guides that contain mixtures of rules where some rules apply to some mime types and other rules apply to different mime types. So putting a mime type qualifier on individual rules seems tedious and surprising to users. If we did allow it (maybe we should for exceptional cases), I think it would only be used in those exceptional cases.

I (again subjectively) don't feel obligated to have resources "all the way down", but even if we did, the user benefit (that I perceive) of easier authoring suggests to me that it's better to keep the mime type list at the style guide level.

This might get clearer when we write some sample style guide YAMLs, like an externalization of the api-linter guidelines, for example.

(I wouldn't think of the style guide as an ancestor to the rules as much as an owner of or context for the rules - so this seems more like a delegation pattern than inheritance)

Reply:
I agree this is a good way to represent the YAML for authoring by users, but I don't think it's a great representation for the application to handle. What's best for users is often not the best internal representation for an application.

>Subjectively, I don't expect that we would see many style guides that contain mixtures of rules where some rules apply to some mime types and other rules apply to different mime types. So putting a mime type qualifier on individual rules seems tedious and surprising to users. If we did allow it (maybe we should for exceptional cases), I think it would only be used in those exceptional cases.

For design flexibility, I think we should consider using rule-level MIME types on the protos as I suggested, but decouple it from the YAML representation which can be kept as-is for ease of authoring. For example, the [server config](https://github.com/apigee/registry/blob/main/cmd/registry-server/main.go) does this, converting from an ideal user representation in YAML to a different internal format. 

This approach doesn't require us to support the "exceptional" cases, but makes it easy to add that functionality if we want to in the future because it's a more flexible (and IMO application-friendly) design. The change would only affect the YAML parsing code instead of everything that touches the protos.

Reply:
OK, I think that's worth considering in the future. These are alpha protos, so we're not committing to them permanently, just as a good starting point.